File: backend/app/execution/utils/parse_date_flexible.py
"""parse_date_flexible — tolerant date parser.

Python mirror of frontend/src/engine/utils/parseDateFlexible.ts. Same
behavior and same test coverage — see backend/tests/test_parse_utilities.py.

Accepts every shape realistic user data produces:
  - ISO 8601: 2026-03-15, 2026/03/15, 2026-03-15T14:30:00
  - dd/mm/yyyy (default)
  - mm/dd/yyyy (when locale_hint='mdy')
  - dd-mm-yyyy
  - Excel serial date (int/float 1..2_958_465)
  - Python datetime / date
  - Month-name forms: "15 Mar 2026", "March 15, 2026"
  - 2-digit years with cutoff (< 50 → 20xx, >= 50 → 19xx)

Returns a `date` on success, `None` on failure.
"""

from __future__ import annotations

import re
from datetime import date, datetime, timedelta
from typing import Any, Literal, Optional

DateLocaleHint = Literal["dmy", "mdy", "auto"]

_MONTH_NAMES: dict[str, int] = {
    "jan": 1, "january": 1,
    "feb": 2, "february": 2,
    "mar": 3, "march": 3,
    "apr": 4, "april": 4,
    "may": 5,
    "jun": 6, "june": 6,
    "jul": 7, "july": 7,
    "aug": 8, "august": 8,
    "sep": 9, "sept": 9, "september": 9,
    "oct": 10, "october": 10,
    "nov": 11, "november": 11,
    "dec": 12, "december": 12,
}

# Excel's epoch quirk: serial 0 = 1899-12-30 in most implementations.
_EXCEL_EPOCH = date(1899, 12, 30)
_EXCEL_MAX_SERIAL = 2_958_465  # 9999-12-31


def _normalize_year(y: int) -> int:
    if y >= 100:
        return y
    return 2000 + y if y < 50 else 1900 + y


def _make_date_if_valid(y: int, m: int, d: int) -> Optional[date]:
    if m < 1 or m > 12 or d < 1 or d > 31:
        return None
    try:
        return date(y, m, d)
    except ValueError:
        return None


def _excel_serial_to_date(serial: float) -> Optional[date]:
    if serial < 1 or serial > _EXCEL_MAX_SERIAL:
        return None
    try:
        return _EXCEL_EPOCH + timedelta(days=int(serial))
    except (OverflowError, ValueError):
        return None


_ISO_RE = re.compile(r"^(\d{4})[-/](\d{1,2})[-/](\d{1,2})(?:[T ].*)?$")
_NUMERIC_DATE_RE = re.compile(r"^(\d{1,4})[-/.](\d{1,2})[-/.](\d{1,4})$")
_DAY_FIRST_MONTHNAME_RE = re.compile(r"^(\d{1,2})[-/ ]+([A-Za-z]+)[-/ ,]+(\d{2,4})$")
_MONTH_FIRST_MONTHNAME_RE = re.compile(r"^([A-Za-z]+)[-/ ]+(\d{1,2})[-/ ,]+(\d{2,4})$")


def parse_date_flexible(
    value: Any,
    locale_hint: DateLocaleHint = "dmy",
) -> Optional[date]:
    # datetime/date pass-through.
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, date):
        return value

    # Excel serial number.
    if isinstance(value, (int, float)) and not isinstance(value, bool):
        return _excel_serial_to_date(float(value))

    if not isinstance(value, str):
        return None

    s = value.strip()
    if not s:
        return None

    # ── ISO: yyyy-mm-dd (unambiguous) ──
    m = _ISO_RE.match(s)
    if m:
        return _make_date_if_valid(int(m.group(1)), int(m.group(2)), int(m.group(3)))

    # ── Month-name forms ──
    m = _DAY_FIRST_MONTHNAME_RE.match(s)
    if m:
        mi = _MONTH_NAMES.get(m.group(2).lower())
        if mi is not None:
            return _make_date_if_valid(_normalize_year(int(m.group(3))), mi, int(m.group(1)))

    m = _MONTH_FIRST_MONTHNAME_RE.match(s)
    if m:
        mi = _MONTH_NAMES.get(m.group(1).lower())
        if mi is not None:
            return _make_date_if_valid(_normalize_year(int(m.group(3))), mi, int(m.group(2)))

    # ── Numeric d-m-y or m-d-y ──
    m = _NUMERIC_DATE_RE.match(s)
    if m:
        a, b, c = int(m.group(1)), int(m.group(2)), int(m.group(3))
        if a >= 1000:
            # Year-leading form we somehow didn't catch above (e.g. dotted).
            return _make_date_if_valid(a, b, c)
        if a > 12 and b <= 12:
            day, month, year = a, b, c
        elif b > 12 and a <= 12:
            day, month, year = b, a, c
        else:
            if locale_hint == "mdy":
                month, day, year = a, b, c
            else:
                day, month, year = a, b, c
        return _make_date_if_valid(_normalize_year(year), month, day)

    return None


def format_date_dmy(d: date) -> str:
    return d.strftime("%d/%m/%Y")


def format_date_mdy(d: date) -> str:
    return d.strftime("%m/%d/%Y")

File: frontend/src/engine/utils/parseDateFlexible.ts
/**
 * parseDateFlexible — tolerant date parser.
 *
 * Accepts every shape that realistic user data can produce:
 *   - ISO 8601:   2026-03-15, 2026-03-15T14:30:00, 2026/03/15
 *   - dd/mm/yyyy: 15/03/2026 (default for most locales)
 *   - mm/dd/yyyy: 03/15/2026 (when localeHint='mdy')
 *   - dd-mm-yyyy: 15-03-2026
 *   - Excel serial date (number 1..2_958_466 — valid Excel date range)
 *   - JavaScript Date object
 *   - Month-name forms:  "15 Mar 2026", "March 15, 2026", "Mar 15 2026"
 *   - 2-digit years with reasonable cutoff (< 50 → 20xx, >= 50 → 19xx)
 *
 * Unambiguous dates (where one number exceeds 12) are parsed correctly
 * regardless of `localeHint`. Ambiguous dates (both ≤ 12) respect
 * `localeHint`; default is `"dmy"` since that is the majority-world format.
 *
 * Returns a UTC `Date` on success, `null` on failure.
 */

export type DateLocaleHint = "dmy" | "mdy" | "auto";

const MONTH_NAMES: Record<string, number> = {
  jan: 0, january: 0,
  feb: 1, february: 1,
  mar: 2, march: 2,
  apr: 3, april: 3,
  may: 4,
  jun: 5, june: 5,
  jul: 6, july: 6,
  aug: 7, august: 7,
  sep: 8, sept: 8, september: 8,
  oct: 9, october: 9,
  nov: 10, november: 10,
  dec: 11, december: 11,
};

function normalizeYear(y: number): number {
  if (y >= 100) return y;
  return y < 50 ? 2000 + y : 1900 + y;
}

/** Validate that (y, m, d) is a real calendar date (rejects Feb 30 etc.). */
function makeDateIfValid(y: number, m: number, d: number): Date | null {
  if (m < 0 || m > 11) return null;
  if (d < 1 || d > 31) return null;
  const date = new Date(Date.UTC(y, m, d));
  if (
    date.getUTCFullYear() !== y ||
    date.getUTCMonth() !== m ||
    date.getUTCDate() !== d
  ) {
    return null;
  }
  return date;
}

/** Excel serial → Date (days since 1899-12-30, accounting for Excel's 1900 leap-year bug). */
function excelSerialToDate(serial: number): Date | null {
  // Excel valid range is ~1900-01-01 (serial 1) through 9999-12-31 (~2_958_465).
  if (!Number.isFinite(serial) || serial < 1 || serial > 2_958_465) return null;
  const excelEpoch = Date.UTC(1899, 11, 30);
  const ms = excelEpoch + serial * 86_400_000;
  const d = new Date(ms);
  return Number.isNaN(d.getTime()) ? null : d;
}

export function parseDateFlexible(
  value: unknown,
  localeHint: DateLocaleHint = "dmy",
): Date | null {
  // Already a Date?
  if (value instanceof Date) {
    return Number.isNaN(value.getTime()) ? null : value;
  }

  // Excel serial (number)?
  if (typeof value === "number") {
    return excelSerialToDate(value);
  }

  if (typeof value !== "string") return null;

  const s = value.trim();
  if (!s) return null;

  // ── ISO: yyyy-mm-dd or yyyy/mm/dd (optionally with Thh:mm:ss — we strip the time part) ──
  // The year-leading form is always unambiguous.
  {
    const m = s.match(/^(\d{4})[-/](\d{1,2})[-/](\d{1,2})(?:[T ].*)?$/);
    if (m) return makeDateIfValid(Number(m[1]), Number(m[2]) - 1, Number(m[3]));
  }

  // ── Month-name forms ──
  // "15 Mar 2026", "15 March 2026", "March 15 2026", "March 15, 2026"
  {
    const dayFirst = s.match(/^(\d{1,2})[-/ ]+([A-Za-z]+)[-/ ,]+(\d{2,4})$/);
    if (dayFirst) {
      const mi = MONTH_NAMES[dayFirst[2].toLowerCase()];
      if (mi !== undefined) {
        return makeDateIfValid(normalizeYear(Number(dayFirst[3])), mi, Number(dayFirst[1]));
      }
    }
    const monthFirst = s.match(/^([A-Za-z]+)[-/ ]+(\d{1,2})[-/ ,]+(\d{2,4})$/);
    if (monthFirst) {
      const mi = MONTH_NAMES[monthFirst[1].toLowerCase()];
      if (mi !== undefined) {
        return makeDateIfValid(normalizeYear(Number(monthFirst[3])), mi, Number(monthFirst[2]));
      }
    }
  }

  // ── Numeric d-m-y or m-d-y with / or - or . separators ──
  {
    const m = s.match(/^(\d{1,4})[-/.](\d{1,2})[-/.](\d{1,4})$/);
    if (m) {
      const a = Number(m[1]);
      const b = Number(m[2]);
      const c = Number(m[3]);

      // Year-leading form like 2026-3-15 should already have matched above.
      // If we reach here with a large first number, treat it defensively as year.
      if (a >= 1000) {
        return makeDateIfValid(a, b - 1, c);
      }

      // Disambiguate: if one number is clearly a day-of-month (>12) we know
      // which slot is the month regardless of locale.
      let day: number, month: number, year: number;
      if (a > 12 && b <= 12) {
        // a is day, b is month → dd/mm/yyyy layout
        day = a; month = b; year = c;
      } else if (b > 12 && a <= 12) {
        // a is month, b is day → mm/dd/yyyy layout
        day = b; month = a; year = c;
      } else {
        // Ambiguous — use locale hint. mdy → a is month; dmy/auto → a is day.
        if (localeHint === "mdy") {
          month = a; day = b; year = c;
        } else {
          day = a; month = b; year = c;
        }
      }
      return makeDateIfValid(normalizeYear(year), month - 1, day);
    }
  }

  // ── Last-ditch: JS Date.parse for oddball formats ──
  // We only accept if the result is a finite Date, to avoid silently returning
  // "Invalid Date".
  const fallback = Date.parse(s);
  if (!Number.isNaN(fallback)) {
    const d = new Date(fallback);
    // Convert to UTC by extracting y/m/d in local time and rebuilding in UTC.
    // Date.parse() interprets yyyy-mm-dd as UTC but "Mar 15 2026" as local.
    // Normalize to a UTC calendar date (time zeroed).
    return new Date(Date.UTC(d.getFullYear(), d.getMonth(), d.getDate()));
  }

  return null;
}

/** Format a UTC Date back to `dd/mm/yyyy` — inverse convention for the dmy locale. */
export function formatDateDMY(d: Date): string {
  const day = String(d.getUTCDate()).padStart(2, "0");
  const mo = String(d.getUTCMonth() + 1).padStart(2, "0");
  const yr = d.getUTCFullYear();
  return `${day}/${mo}/${yr}`;
}

/** Format a UTC Date back to `mm/dd/yyyy`. */
export function formatDateMDY(d: Date): string {
  const day = String(d.getUTCDate()).padStart(2, "0");
  const mo = String(d.getUTCMonth() + 1).padStart(2, "0");
  const yr = d.getUTCFullYear();
  return `${mo}/${day}/${yr}`;
}
